api/src/main/java/org/openmrs/module/htmlformentryui/element/ObsFromFragmentElement.java

```diff
@@ -0,0 +1,328 @@
+import org.openmrs.Concept;
+import org.openmrs.ConceptAnswer;
+import org.openmrs.ConceptDatatype;
+import org.openmrs.Encounter;
+import org.openmrs.Obs;
+import org.openmrs.module.htmlformentry.FormEntryContext;
+import org.openmrs.module.htmlformentry.FormEntrySession;
+import org.openmrs.module.htmlformentry.FormSubmissionError;
+import org.openmrs.module.htmlformentry.HtmlFormEntryUtil;
+import org.openmrs.module.htmlformentry.action.FormSubmissionControllerAction;
+import org.openmrs.module.htmlformentry.element.HtmlGeneratorElement;
+import org.openmrs.module.htmlformentryui.tag.UiIncludeTagHandler;
+import org.openmrs.ui.framework.FragmentException;
+import org.openmrs.ui.framework.UiUtils;
+import org.openmrs.ui.framework.page.PageAction;
+
+public class ObsFromFragmentElement implements HtmlGeneratorElement, FormSubmissionControllerAction {
+
+	private UiUtils uiUtils;
+	private Map<String, Object> fragmentParams;
+	private Concept concept;
+	private String viewProvider;
+	private String fragment;
+	private String initFragmentParamName;
+	
+	// For tests only
+	public ObsFromFragmentElement() {
+		
+	}
+	
+	public ObsFromFragmentElement(Map<String, String> parameters, UiUtils uiUtils) {
+		this.uiUtils = uiUtils;
+		concept = HtmlFormEntryUtil.getConcept(parameters.get("conceptId"));
+		if (concept == null) {
+			throw new IllegalArgumentException("Cannot find concept with id " + parameters.get("conceptId"));
+		}
+		viewProvider = parameters.get("provider");
+		if (StringUtils.isBlank(viewProvider)) {
+			throw new IllegalArgumentException("Parameter provider cannot be blank");
+		}
+		fragment = parameters.get("fragment");
+		if (StringUtils.isBlank(fragment)) {
+			throw new IllegalArgumentException("Parameter fragment cannot be blank");
+		}
+		initFragmentParamName = parameters.get("initFragmentParamName");
+		if (StringUtils.isBlank(initFragmentParamName)) {
+			throw new IllegalArgumentException("Parameter initFragmentParamName cannot be blank");
+		}
+		StringBuilder fragmentUrlBuilder = new StringBuilder();
+		fragmentUrlBuilder.append(fragment);
+		fragmentUrlBuilder.append("?");
+		fragmentUrlBuilder.append(parameters.get("fragmentParams"));
+		try {
+			fragmentParams = UiIncludeTagHandler.paramsToMap(fragmentUrlBuilder.toString());
+							
+		} catch (URISyntaxException e) {
+			throw new FragmentException("Invalid fragment URI: " + fragmentUrlBuilder.toString() , e);
+		}
+	}
+	
+	@Override
+	public void handleSubmission(FormEntrySession session, HttpServletRequest request) {
+		String formFieldName = (String) fragmentParams.get("formFieldName");
+		if (formFieldName != null) {
+			String valueText = request.getParameter(formFieldName);
+			if (session.getContext().getMode().equals(FormEntryContext.Mode.EDIT)) {
+				List<Obs> existingObservations = session.getContext().removeExistingObs(concept);
+				if (existingObservations != null) {
+					if (!existingObservations.isEmpty()) {
```

Comment:
Instead of those nested `if`, did you mean `!CollectionUtils.isEmpty(existingObservations)`?

Reply:
yeah sure